down-the-stack/dk_template/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+//! Hardware Abstraction Layer (HAL) for the nRF52840 Development Kit
+
+#![deny(missing_docs)]
+#![deny(warnings)]
+#![no_std]
+
+use core::{
+    ops,
+    fmt,
+    sync::atomic::{self, AtomicU32, Ordering},
+    time::Duration,
+};
+
+use cortex_m::{asm, peripheral::NVIC};
+use embedded_hal::digital::v2::{OutputPin as _, StatefulOutputPin};
+pub use hal::ieee802154;
+pub use hal::pac::{
+    interrupt, Interrupt, NVIC_PRIO_BITS, RTC0, UARTE1, uarte0::{
+    baudrate::BAUDRATE_A as Baudrate, config::PARITY_A as Parity}};
+use hal::{
+    clocks::{self, Clocks},
+    gpio::{p0, Level, Output, Input, PullUp, Pin, Port, PushPull},
+    rtc::{Rtc, RtcInterrupt},
+    timer::OneShot, prelude::InputPin,
+};
+
+use defmt;
+use defmt_rtt as _; // global logger
+
+
+use crate::{
+    peripheral::{POWER, USBD},
+    usbd::Ep0In,
+ 
+};
+
+mod errata;
+pub mod peripheral;
+pub mod usbd;
+
+
+/// Components on the board
+pub struct Board {
+    /// LEDs
+    pub leds: Leds,
+    // --- Exercise --- 🔽
+    /// Buttons
+    pub buttons: Buttons,
+    // --- Exercise --- 🔼 
+    /// Timer
+    pub timer: Timer,
+
+    /// Radio interface
+    pub radio: ieee802154::Radio<'static>,
+    /// USBD (Universal Serial Bus Device) peripheral
+    pub usbd: USBD,
+    /// POWER (Power Supply) peripheral
+    pub power: POWER,
+    /// USB control endpoint 0
+    pub ep0in: Ep0In,
+    // --- Exercise --- 🔽
+    /// uarte interface
+    pub uarte: Uarte,
+    // --- Exercise --- 🔼 
+}
+
+/// All LEDs on the board
+pub struct Leds {
+    /// LED1: pin P0.13, green LED
+    pub led_1: Led,
+    /// LED2: pin P0.14, green LED
+    pub led_2: Led,
+    /// LED3: pin P0.15, green LED
+    pub led_3: Led,
+    /// LED4: pin P0.16, green LED
+    pub led_4: Led,
+}
+
+/// A single LED
+pub struct Led {
+    inner: Pin<Output<PushPull>>,
+}
+
+impl Led {
+    /// Turns on the LED
+    pub fn on(&mut self) {
+        defmt::trace!(
+            "setting P{}.{} low (LED on)",
+            if self.inner.port() == Port::Port1 {
```

Comment:
Maybe move this off into a helper function?

```rust
fn port_as_char(port: &Port) -> char {
    match port {
        Port::Port0 => '0',
        Port::Port1 => '1',
    }
}
```

This will also helpfully break when someone makes an nRF chip with more than two ports.